ygen/goelements_test.go

```diff
@@ -1404,3 +1410,774 @@ func TestTypeResolutionManyToOne(t *testing.T) {
+		ctx: &yang.Entry{
+			Name: "union-leaf",
+			Kind: yang.LeafEntry,
+			Type: &yang.YangType{
+				Name: "union",
+				Kind: yang.Yunion,
+				Type: []*yang.YangType{{
+					Kind:    yang.Yidentityref,
+					Name:    "identityref",
+					Default: "prefix:CHIPS",
+					IdentityBase: &yang.Identity{
+						Name: "base-identity",
+						Parent: &yang.Module{
+							Name: "base-module",
+						},
+						Values: []*yang.Identity{
+							{Name: "FOO"},
+							{Name: "BAR"},
+						},
+					},
+				}, {
+					Kind: yang.Ystring,
+				}},
+			},
+			Parent: &yang.Entry{Name: "base-module"},
+			Node: &yang.Leaf{
+				Parent: &yang.Module{
+					Name: "base-module",
+				},
+			},
+		},
+		in:       "BOO",
+		want:     ygot.String("UnionString(\"BOO\")"),
+		wantKind: yang.Yunion,
+	}, {
+		name: "enumeration",
+		ctx: &yang.Entry{
+			Name: "enumeration-leaf",
+			Type: &yang.YangType{
+				Name: "enumeration",
+				Kind: yang.Yenum,
+				Enum: testEnumType,
+			},
+			Parent: &yang.Entry{Name: "base-module"},
+			Node: &yang.Identity{
+				Parent: &yang.Module{Name: "base-module"},
+			},
+		},
+		in:       "BLUE",
+		want:     ygot.String("BaseModule_EnumerationLeaf_BLUE"),
+		wantKind: yang.Yenum,
+	}, {
+		name: "enumeration not found",
+		ctx: &yang.Entry{
+			Name: "enumeration-leaf",
+			Type: &yang.YangType{
+				Name: "enumeration",
+				Kind: yang.Yenum,
+				Enum: testEnumType,
+			},
+			Parent: &yang.Entry{Name: "base-module"},
+			Node: &yang.Identity{
+				Parent: &yang.Module{Name: "base-module"},
+			},
+		},
+		in:       "GREEN",
+		wantErr:  true,
+		wantKind: yang.Ynone,
+	}, {
+		name: "enumeration in union as the lone type, with prefix",
+		ctx: &yang.Entry{
+			Name: "union-leaf",
+			Kind: yang.LeafEntry,
+			Type: &yang.YangType{
+				Name: "union",
+				Kind: yang.Yunion,
+				Type: []*yang.YangType{
+					{
+						Name: "enumeration",
+						Kind: yang.Yenum,
+						Enum: testEnumType,
+					},
+				},
+			},
+			Parent: &yang.Entry{Name: "base-module"},
+			Node: &yang.Enum{
+				Name:   "enum",
+				Parent: &yang.Module{Name: "base-module"},
+			},
+		},
+		in:       "oc:BLUE",
+		want:     ygot.String("BaseModule_UnionLeaf_Enum_BLUE"),
+		wantKind: yang.Yunion,
+	}, {
+		name: "enumeration in union with string as the second union type",
+		ctx: &yang.Entry{
+			Name: "union-leaf",
+			Kind: yang.LeafEntry,
+			Type: &yang.YangType{
+				Name: "union",
+				Kind: yang.Yunion,
+				Type: []*yang.YangType{{
+					Name: "enumeration",
+					Kind: yang.Yenum,
+					Enum: testEnumType,
+				}, {
+					Kind: yang.Ystring,
+				}},
+			},
+			Parent: &yang.Entry{Name: "base-module"},
+			Node: &yang.Enum{
+				Name:   "enum",
+				Parent: &yang.Module{Name: "base-module"},
+			},
+		},
+		in:       "BLUE",
+		want:     ygot.String("BaseModule_UnionLeaf_Enum_BLUE"),
+		wantKind: yang.Yunion,
+	}, {
+		name: "enumeration in union with string as the first union type",
```

Comment:
I don't think this is the right behaviour -- if a union is:

```yang
type union { 
  type string;
  type enumeration { enum BLUE; }
}
default "BLUE";
```

then the default should be the string `BLUE`. The order in the union matters rather than the most specific type that can be matched according to my understanding.

Reply:
I'll change it, I see now that we're going forward with using the right behaviour in 1.0.0 (https://github.com/openconfig/ygot/issues/430). It makes sense to do it right the first time even if it doesn't conform with the current unmarshalling code.

Opened https://github.com/openconfig/models/pull/924 to fix the bug in openconfig-macsec.yang.

Reply:
Actually, there may not be a bug in macsec after all, since date-and-time has a regex... I'll incorporate the regex check in the conversion code.

Reply:
I've changed the implementation to check in order of union subtype definition.